src/CBLCollection_Internal.hh

```diff
@@ -194,17 +179,23 @@ private:
 #pragma mark - LISTENERS:
     
     Retained<CBLListenerToken> addListener(fleece::function_ref<Retained<CBLListenerToken>()> cb) {
-        auto lock = database()->useLocked();
         Retained<CBLListenerToken> token = cb();
         if (!_observer)
-            _observer = _c4col->observe([this](C4CollectionObserver*) { this->collectionChanged(); });
+            _observer = _c4col.useLocked()->observe([this](C4CollectionObserver*) {
+                this->collectionChanged();
+            });
         return token;
     }
     
     void collectionChanged() {
+        Retained<CBLDatabase> db;
         try {
-            database()->notify(std::bind(&CBLCollection::callCollectionChangeListeners, this));
+            db = database();
         } catch (...) { }
```

Comment:
Log here too.  I can just imagine a situation where a user is claiming that they receive no notifications, and the logs could tell us here what is happening.

Reply:
Added.